### PR TITLE
[LoongArch64] add the coreclr/pal directory

### PR DESCRIPTION
**Pull request by shushanhf**
[LoongArch64] add the coreclr/pal and libunwind directory.

### CODE REVIEW

**Comment by am11** [comment]
@shushanhf, I assume you have just applied https://github.com/libunwind/libunwind/commit/c5f1d12c77dea6a60740730c675fc56b3c52b86a on v1.6.2 and not copied the upstream master branch, right? If not please only apply the commit which is needed. Also, please update the version file so we can keep track of changes. This helps us when updating libunwind dependency. e.g. in `src/coreclr/pal/src/libunwind/libunwind-version.txt`, add a line:

```
Apply https://github.com/libunwind/libunwind/pull/316
```

**Comment by shushanhf** [comment]
> make

In fact we used the first version libunwind for .NET6 and passed, then our company commits it to the upstream.

**Comment by shushanhf** [comment]
> @shushanhf, I assume you have just applied [libunwind/libunwind@c5f1d12](https://github.com/libunwind/libunwind/commit/c5f1d12c77dea6a60740730c675fc56b3c52b86a) on v1.6.2 and not copied the upstream master branch, right? If not please only apply the commit which is needed. Also, please update the version file so we can keep track of changes. This helps us when updating libunwind dependency. e.g. in `src/coreclr/pal/src/libunwind/libunwind-version.txt`, add a line:
> 
> ```
> Apply https://github.com/libunwind/libunwind/pull/316
> ```

**Comment by am11** [comment]
I think if you could create a separate PR just for libunwind, as @jkotas suggested (collecting only the libunwind related changes from https://github.com/dotnet/arcade/pull/8271 and this PR), i.e. everything under `src/coreclr/pal/src/libunwind` and nothing outside, that would make it quicker to review. Here is an example of such a PR for s390x architecture: https://github.com/dotnet/runtime/pull/53286.

In terms of bash:
```sh
# start fresh
git remote add dotnet https://github.com/dotnet/runtime
git checkout -b feature/loongarch64/libunwind
git fetch --all
git reset --hard dotnet/main

# apply a clean patch
curl -sSL https://github.com/libunwind/libunwind/commit/c5f1d12.patch |\
    git apply - --directory src/coreclr/pal/src/libunwind

# manually update src/coreclr/pal/src/libunwind/libunwind-version.txt
# manually update CMakeLists.txt etc. (as you did in this PR: #62092)
git commit -m "Add loongarch64 support in libunwind"
git push -u origin feature/loongarch64/libunwind
# create a new PR from branch 'feature/loongarch64/libunwind'
```
The next time someone will update this lib, it will save them trouble to compare files and instead they will just follow our regular steps (https://github.com/dotnet/runtime/pull/62092#issue-1064865913). I am working on improving it a little bit so we don't modify CMakeLists.txt and friends in vendor's directory and changing libunwind's location to `src/native/external`; https://github.com/dotnet/runtime/compare/main...am11:feature/consolidations, it is actually ready, I'm just waiting for yours loongarch64 changes to go in first to avoid merge conflicts. :)

**Comment by shushanhf** [comment]
> I think if you could create a separate PR just for libunwind, as @jkotas suggested (collecting only the libunwind related changes from #62889 and this PR), i.e. everything under `src/coreclr/pal/src/libunwind` and nothing outside, that would make it quicker to review. Here is an example of such a PR for s390x architecture: #53286.
> 
> In terms of bash:
> 
> ```shell
> # start fresh
> git remote add dotnet https://github.com/dotnet/runtime
> git checkout -b feature/loongarch64/libunwind
> git fetch --all
> git reset --hard dotnet/main
> 
> # apply a clean patch
> curl -sSL https://github.com/libunwind/libunwind/commit/c5f1d12.patch |\
>     git apply - --directory src/coreclr/pal/src/libunwind
> 
> # manually update src/coreclr/pal/src/libunwind/libunwind-version.txt
> # manually update CMakeLists.txt etc. (as you did in this PR: #62092)
> git commit -m "Add loongarch64 support in libunwind"
> git push -u origin feature/loongarch64/libunwind
> # create a new PR from branch 'feature/loongarch64/libunwind'
> ```
> 
> The next time someone will update this lib, it will save them trouble to compare files and instead they will just follow our regular steps ([#62092 (comment)](https://github.com/dotnet/runtime/pull/62092#issue-1064865913)). I am working on improving it a little bit so we don't modify CMakeLists.txt and friends in vendor's directory and changing libunwind's location to `src/native/external`; [main...am11:feature/consolidations](https://github.com/dotnet/runtime/compare/main...am11:feature/consolidations), it is actually ready, I'm just waiting for yours loongarch64 changes to go in first to avoid merge conflicts. :)

Thanks for your review and advices,
I will update it later.

**Comment by shushanhf** [comment]
> I think if you could create a separate PR just for libunwind, as @jkotas suggested (collecting only the libunwind related changes from #62889 and this PR), i.e. everything under `src/coreclr/pal/src/libunwind` and nothing outside, that would make it quicker to review. Here is an example of such a PR for s390x architecture: #53286.
> 
> In terms of bash:
> 
> ```shell
> # start fresh
> git remote add dotnet https://github.com/dotnet/runtime
> git checkout -b feature/loongarch64/libunwind
> git fetch --all
> git reset --hard dotnet/main
> 
> # apply a clean patch
> curl -sSL https://github.com/libunwind/libunwind/commit/c5f1d12.patch |\
>     git apply - --directory src/coreclr/pal/src/libunwind
> 
> # manually update src/coreclr/pal/src/libunwind/libunwind-version.txt
> # manually update CMakeLists.txt etc. (as you did in this PR: #62092)
> git commit -m "Add loongarch64 support in libunwind"
> git push -u origin feature/loongarch64/libunwind
> # create a new PR from branch 'feature/loongarch64/libunwind'
> ```
> 
> The next time someone will update this lib, it will save them trouble to compare files and instead they will just follow our regular steps ([#62092 (comment)](https://github.com/dotnet/runtime/pull/62092#issue-1064865913)). I am working on improving it a little bit so we don't modify CMakeLists.txt and friends in vendor's directory and changing libunwind's location to `src/native/external`; [main...am11:feature/consolidations](https://github.com/dotnet/runtime/compare/main...am11:feature/consolidations), it is actually ready, I'm just waiting for yours loongarch64 changes to go in first to avoid merge conflicts. :)

Had pushed the libunwind separately by #62979

File: src/coreclr/pal/src/eventprovider/lttngprovider/CMakeLists.txt [hunk 1]
```diff
@@ -62,9 +62,20 @@ add_dependencies(coreclrtraceptprovider generated_eventing_headers)
...
+else()
+  target_link_libraries(coreclrtraceptprovider
+                        lttng-ust-tracepoint
+                        urcu
+                        urcu-bp
+                        urcu-cds
+                        urcu-common
+                        ${LTTNG}
+  )
+endif()
```
Comment (am11) [inline src/coreclr/pal/src/eventprovider/lttngprovider/CMakeLists.txt:78]: Curious, what is the error if we don't make this change? lttng-ust-tracepoint is actually a soft dependency, i.e. this library gets loaded via `dlopen()` during the run time of application. Even the existing `target_link_libraries` is not needed, it is a leftover from past refactoring, we can just delete it altogether.

Comment (shushanhf) [reply src/coreclr/pal/src/eventprovider/lttngprovider/CMakeLists.txt:78]: > Curious, what is the error if we don't make this change? lttng-ust-tracepoint is actually a soft dependency, i.e. this library gets loaded via `dlopen()` during the run time of application. Even the existing `target_link_libraries` is not needed, it is a leftover from past refactoring, we can just delete it altogether.

Yes, I thinks so.
I check it within my machine now.

Comment (shushanhf) [reply src/coreclr/pal/src/eventprovider/lttngprovider/CMakeLists.txt]: Thanks for your review!
I had deleted the modification.  (I guess that it's old code from the mips64 when we porting it.)

**Comment by shushanhf** [comment]
Now this PR can be compiled sucessfully on the native mode.
For example,
https://dev.azure.com/dnceng/public/_build/results?buildId=1522042&view=logs&j=86357178-b8b8-53d0-d37b-335a1e08c946&t=75d29754-3c44-5c16-fcdb-e0edbf440620

What should I do next ?

**Comment by shushanhf** [comment]
Now the compiling errors are within the manged product.
Like this:
https://dev.azure.com/dnceng/public/_build/results?buildId=1522199&view=logs&j=11a68750-4cc0-50f6-a992-7042650640b8&t=26e8beda-8bda-5022-b907-3e9f228434d9

**Comment by am11** [comment]
The native and managed jitinterfaces used by crossgen2 are not in synchronized state anymore. If `int *flags` argument is needed by loongarch64, maybe add it to the existing `getArgType` and leave it unused on the other architectures (`#ifndef TARGET_LOONGARCH64\n(void)flags; // unused`). After that run `src/coreclr/tools/Common/JitInterface/ThunkGenerator/gen.sh` from root of the repo which will take a minute and synchronize the native and managed sides of JIT interface, and adjust the list ordinals in `CorInfoBase.cs`.

**Comment by jkotas** [comment]
The JIT interface changes should be excluded from this PR. This PR should be only touching PR.

**Comment by am11** [comment]
Agreed with @jkotas. PAL changes in a separate PR as a first component during a platform port makes things much clearer for coreclr partition.

@shushanhf, could you please revert changes in files which are not under `src/coreclr/pal` directory? Also:
* remove libunwind from this PR's title
* build and run PAL tests to make sure this base layer is working as expected on linux loongarch64: https://github.com/dotnet/runtime/blob/3b40c7c/docs/workflow/testing/coreclr/unix-test-instructions.md#pal-tests

**Comment by shushanhf** [comment]
> The JIT interface changes should be excluded from this PR. This PR should be only touching the PAL.



> Agreed with @jkotas. PAL changes in a separate PR as a first component during a platform port makes things much clearer for coreclr partition.
> 
> @shushanhf, could you please revert changes in files which are not under `src/coreclr/pal` directory? Also:
> 
> * remove libunwind from this PR's title
> * build and run PAL tests to make sure this base layer is working as expected on linux loongarch64: https://github.com/dotnet/runtime/blob/3b40c7c/docs/workflow/testing/coreclr/unix-test-instructions.md#pal-tests

Thanks, I will do it.

**Comment by BruceForstall** [comment]
@janvorli Is this something you could review?
cc @mangod9 

@shushanhf Looks like some jobs failed. Maybe you need to make sure the branch is up to date and retrigger the testing.

**Comment by janvorli** [comment]
> @janvorli Is this something you could review?

Yes, I will take a look today.

File: src/coreclr/pal/src/thread/context.cpp [hunk 2]
```diff
@@ -227,7 +268,11 @@ BOOL CONTEXT_GetRegisters(DWORD processId, LPCONTEXT lpContext)
 
     if (processId == GetCurrentProcessId())
     {
+#ifndef HOST_LOONGARCH64
         CONTEXT_CaptureContext(lpContext);
+#else
+        _ASSERT(!"unimplemented on LOONGARCH64 yet");
```
Comment (janvorli) [inline src/coreclr/pal/src/thread/context.cpp:274]: Should this be implemented in this PR?

File: src/coreclr/pal/inc/pal.h [hunk 3]
```diff
@@ -2453,6 +2601,9 @@ PALIMPORT BOOL PALAPI PAL_VirtualUnwindOutOfProc(CONTEXT *context, KNONVOLATILE_
 #define PAL_CS_NATIVE_DATA_SIZE 56
 #elif defined(__sun) && defined(__x86_64__)
 #define PAL_CS_NATIVE_DATA_SIZE 48
+#elif defined(__linux__) && defined(__loongarch64)
+////TODO for LOONGARCH64: should confirm !!!
```
Comment (janvorli) [inline src/coreclr/pal/inc/pal.h:2605]: It would be good to confirm this before merging this PR.

**Review by janvorli — COMMENTED** [review]
Overall it looks good, I have just two comments.

File: src/coreclr/pal/src/thread/context.cpp (hunk 2 above)
Comment (shushanhf) [reply src/coreclr/pal/src/thread/context.cpp]: OK.
I had delete the assert for having implements of the `CONTEXT_CaptureContext`.

File: src/coreclr/pal/inc/pal.h (hunk 3 above)
Comment (shushanhf) [reply src/coreclr/pal/inc/pal.h]: Thanks.
I have update it.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Comment by BruceForstall** [comment]
The CI failures look like unrelated infrastructure. Merging.

**Pull request merged**